### PR TITLE
Docs: guide copy edits, subjective improvements, and other suggestions

### PR DESCRIPTION
**Pull request by ches**
Greetings Rustaceans!

I've just been getting acquainted with Rust through the guide. First let me say that it's already in great shape, chapters are kept a good length to be digestible and paced to move the reader along fluidly, so my compliments to contributors!

Along the way I noticed a few minor copy errors, and also a few areas that I thought more subjectively could stand to be improved. My commits here are divided so that minor edits unlikely to be very contentious could be cherry-picked, and then topically on parts that might generate more discussion.

I also have some comments and questions that aren't directly associated with any changes on this branch yet. I'm not sure how you like to triage this sort of thing but I'll present them below and if it's appropriate they could be moved to separate issues or I might be able to help work some of them out within the scope of this PR. Sorry that these are a lot to take in, pretty much everything below here can be digested independently of the current changes in this PR so you could read the rest later :smile:
### Questions and Comments

I'll give stable links to doc revisions as of this writing.
1. The [example using `PartialEq` in the Traits chapter](https://github.com/rust-lang/rust/blob/157614249594f187f421cd97f928e64c5ab5c1fa/src/doc/trpl/traits.md#our-inverse-example) is poor—we have no idea how `PartialEq` works at this point in the text (or at any point, AFAICT), so it isn't clear why it won't work as a trait bound in this situation and `Float` almost magically does, with the aid of existing tailor-made identity functions that seem unlikely to be so conveniently available when we encounter a scenario like this in our real-world code.
   
   This section just seems glossed over, or perhaps content has moved around over time or there's an assumption that implementing equality with `PartialEq` should be covered in the guide eventually so this example will be less foreign. As it stands the text is hard to follow and not very meaningful.
2. I found treatment of the relationship of trait objects to pointers in the _Static and Dynamic Dispatch_ chapter unclear. [The "Why Pointers?" section](https://github.com/rust-lang/rust/blob/157614249594f187f421cd97f928e64c5ab5c1fa/src/doc/trpl/static-and-dynamic-dispatch.md#why-pointers) opens with this line:
   
   > The use of language like "fat pointer" implies that a trait object is always a pointer of some form, but why?
   
   But the phrase "fat pointer" hasn't been used anywhere before. This is some of the more complex material in the guide, but this section nevertheless feels displaced, not clearly connecting preceding subject matter. Earlier we've covered the internal representation of trait objects and significance of pointers they contain, but it hasn't been spelled out (other than what `&Foo` syntax suggests) that trait objects are references (and why). That's what the "Why Pointers?" section is aiming to do I gather, but it seems out of place, I think it'd make more sense to cover this before the gory details of their internals.
3. Suggestion: move the _Error Handling_ chapter much earlier in the Intermediate section of the guide, or even into the Basics section. I know the Intermediate section isn't intended to be read in order per se, but plenty of people like me are just going to read it straight through anyway :grin: These are pretty fundamental concepts to understand and `Option`, `Result`, and idioms like `unwrap()` and `.ok().expect()` are referenced numerous times throughout the rest of the guide. They feature pretty prominently as early as _Standard Input_ and _Guessing Game_ chapters in Basics, in fact. I happen to have a good understanding of these already through encountering their analogs in typed functional languages, but if I didn't I believe I really would have appreciated reading _Error Handling_ much earlier.
4. In the `rustdoc` chapter, a [comment at the beginning of the first source example](https://github.com/rust-lang/rust/blob/157614249594f187f421cd97f928e64c5ab5c1fa/src/doc/trpl/documentation.md#creating-documentation) refers to a "link" crate attribute being needed. There seems to be no such attribute present in the source. I believe this refers to `crate_type` [according to the reference](http://doc.rust-lang.org/reference.html#linkage), but it'd be nice if this example were updated/clarified (I think `crate_id` is deprecated/obsolete too).
   
   This brings me to a related comment also: after encountering crate attributes in the reference and also docs on Cargo configuration like `crate-type = ["dylib"]`, I'm uncertain about the relationship/redundancy between these. I'm sure this is the kind of thing where docs are simply struggling to keep pace with rapid changes in Rust and Cargo, just wanted to flag that this distinction ought to be clearly covered in the docs for one or the other at some point, it's presently hard to track down.
5. Minor: link to sample editor configurations in [the introductory chapter](https://github.com/rust-lang/rust/blob/157614249594f187f421cd97f928e64c5ab5c1fa/src/doc/trpl/hello-world.md) is broken, probably the generator automatically translates `.md` links to `.html`. Perhaps it shouldn't do that for absolute URLs.
6. Following from my changes to the enums coverage in [_Compound Data Types_](https://github.com/rust-lang/rust/blob/157614249594f187f421cd97f928e64c5ab5c1fa/src/doc/trpl/compound-data-types.md#enums) in this PR: sum types are an important topic and I tried to make some improvements, but I think the motivating example of `Character` with `Digit(i32)` and `Other` variants is a pretty weak one, and a better example could greatly improve cohesion with the `Ordering` coverage later in the section and how that ties into pattern matching in the subsequent chapter. I just haven't thought of a better example to suggest yet.
   
   In particular, the text states:
   
   > This may seem rather limiting, but it's a limitation which we can overcome.
   
   This is referring to `Character`, and actually to more than one limitation: the preceding admonition that its variants aren't comparable/don't have ordering, and don't support binary operators like `*` and `+`. Overcoming these limitations actually never gets explained—we next cover how `Ordering` works as an enum itself for plain `i32`s, but never get around to showing how this might be applied to our `Digit` variant type.
   
   Since the coverage of enums already segues into pattern matching and this could be even tighter with a stronger example, it might be nice if our example enum were somehow connected to the final example program for the Basics section too, where `Ordering` reappears. I don't see how it would fit with the current guessing game example, but food for thought.
7. `#[derive]` seems conspicuously missing from the guide. It would probably make sense to introduce after showing simple examples of implementing equality and/or ordering traits by hand, which have been mentioned as possibilities above. Perhaps it's too much to breach this as early as the Basic section though without traits being introduced. `#[derive]` itself and the derivable traits can certainly be saved for Intermediate and referenced as covered later, in any case.

r? @steveklabnik for docs.

### CODE REVIEW

**Comment by steveklabnik** [comment]
Hey @ches!

First of all, thank you for putting this level of care into a PR. I haven't reviewed the diff yet, but I love getting huge descriptions.

That being said, it's also very hard to manage big ones with unrelated changes like this ,so feel free to open individual ones in the future, as they tend to be much easier to handle, since we can handle them peacemeal. The cherry-picking thing is true, but since @bors can't do that automatically, in our case, it's a significantly more work.

On to the diff!

**Comment by steveklabnik** [comment]
Broad overview comment: much of what you're talking about is basic fallout from moving from "The guide + guides" to a single, unified book, and it hasn't gotten a full editing pass yet, becuase I'm still writing some chapters. So good eyes, and I agree.

**Comment by steveklabnik** [comment]
The Rustdoc chapter is being entirely re-written in https://github.com/rust-lang/rust/pull/22549, so that will be fixed by that.

**Comment by steveklabnik** [comment]
> #[derive] seems conspicuously missing from the guide.

There's still a lot of things missing, but you're right. It would belong in the trait chapter, for sure.

File: src/doc/trpl/arrays-vectors-and-slices.md [hunk 1]
```diff
@@ -49,7 +49,7 @@ languages.
 
 A *vector* is a dynamic or "growable" array, implemented as the standard
 library type [`Vec<T>`](../std/vec/) (we'll talk about what the `<T>` means
-later). Vectors always allocate their data on the heap. Vectors are to slices 
+later). Vectors always allocate their data on the heap. Vectors are to arrays
```
Comment (steveklabnik) [inline src/doc/trpl/arrays-vectors-and-slices.md]: This is wrong, the original is correct.

File: src/doc/trpl/compound-data-types.md [hunk 2]
```diff
@@ -47,8 +47,8 @@ This pattern is very powerful, and we'll see it repeated more later.
 
 There are also a few things you can do with a tuple as a whole, without
 destructuring. You can assign one tuple into another, if they have the same
-contained types and arity. Tuples have the same arity when they have the same
-length.
+contained types and [arity](glossary.html#arity). Tuples have the same arity
+when they have the same length.
```
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: Ah ha, nice catch!

File: src/doc/trpl/compound-data-types.md [hunk 3]
```diff
@@ -196,8 +196,8 @@ Now, we have actual names, rather than positions. Good names are important,
 and with a struct, we have actual names.
 
 There _is_ one case when a tuple struct is very useful, though, and that's a
-tuple struct with only one element. We call this a *newtype*, because it lets
-you create a new type that's similar to another one:
+tuple struct with only one element. We call this the *newtype* pattern, because
+it lets you create a new type that adds semantic meaning to another one:
```
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: It's not just additive, though, as newtypes don't inherit all of the behavior of the old one, it's an entirely new thing.

File: src/doc/trpl/compound-data-types.md [hunk 4]
```diff
@@ -228,8 +228,8 @@ enum Character {
 }
 ```
 
-An `enum` variant can be defined as most normal types. Below are some example
-types have been listed which also would be allowed in an `enum`.
+Most normal types are allowed as the variant components of an `enum`. Below are
+some examples.
```
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: What about "Here are some examples:" which feels a bit more consistent with other places. I may just be remembering the frequency of stuff wrong though...

File: src/doc/trpl/compound-data-types.md [hunk 5]
```diff
@@ -347,14 +347,15 @@ fn main() {
...
-As you can see, `enum`s are quite a powerful tool for data representation, and are
-even more useful when they're [generic][generics] across types. Before we
-get to generics, though, let's talk about how to use them with pattern matching, a
-tool that will let us deconstruct this sum type (the type theory term for enums)
-in a very elegant way and avoid all these messy `if`/`else`s.
+so do this with caution. For this reason, it's normally considered better style
+to `use` an enum rather than its variants directly.
+
+As you can see, `enum`s are quite a powerful tool for data representation, and
+are even more useful when they're [generic][generics] across types. Before we
```
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: Is this link defined anywhere?

File: src/doc/trpl/guessing-game.md [hunk 6]
```diff
@@ -95,8 +95,9 @@ import the `std::old_io` module with `use`, and then our `main` function contain
 our program's logic. We print a little message announcing the game, ask the
 user to input a guess, get their input, and then print it out.
 
-Because we talked about this in the section on standard I/O, I won't go into
-more details here. If you need a refresher, go re-read that section.
+Because we talked about this in [the section on standard
+I/O](standard-input.html), I won't go into more details here. If you need a
+refresher, go re-read that section.
```
Comment (steveklabnik) [inline src/doc/trpl/guessing-game.md]: We've been preferring the `[][]` style links, can we add it with that instead of inline?

File: src/doc/trpl/match.md [hunk 7]
```diff
@@ -20,25 +20,26 @@ match x {
 }
 ```
 
-`match` takes an expression and then branches based on its value. Each *arm* of
-the branch is of the form `val => expression`. When the value matches, that arm's
+`match` takes an expression and then branches based on its value. Each branch or
+*case* is of the form `val => expression`. When the value matches, that case's
```
Comment (steveklabnik) [inline src/doc/trpl/match.md]: We do call them arms, not cases.

File: src/doc/trpl/standard-input.md [hunk 8]
```diff
@@ -115,8 +115,9 @@ doesn't work, so we're okay with that. In most cases, we would want to handle
 the error case explicitly. `expect()` allows us to give an error message if
 this crash happens.
 
-We will cover the exact details of how all of this works later in the Guide.
-For now, this gives you enough of a basic understanding to work with.
+We will cover the exact details of how all of this works later in the Guide in
+[Error Handling](error-handling.html). For now, this gives you enough of a
```
Comment (steveklabnik) [inline src/doc/trpl/standard-input.md]: same here with `[][]`.

**Comment by steveklabnik** [comment]
As far as the diff goes, I'm generally happy with all of these changes, thanks. I left some nits inline :)

**Comment by ches** [comment]
> That being said, it's also very hard to manage big ones with unrelated changes like this ,so feel free to open individual ones in the future, as they tend to be much easier to handle, since we can handle them peacemeal. The cherry-picking thing is true, but since @bors can't do that automatically, in our case, it's a significantly more work.

Noted. With source it's usually pretty obvious how to limit the scope of a PR, with docs I'd assume it's easier to review at once, but I won't make that assumption if I have doc changes again. Not used to the workflow of the bot, because otherwise I don't know how I would've separated the commits any differently really.

Anyhoo.

File: src/doc/trpl/arrays-vectors-and-slices.md (hunk 1 above)
Comment (ches) [inline src/doc/trpl/arrays-vectors-and-slices.md]: I pondered this twice on two separate passes and still decided my change was correct, so maybe that says something about ambiguity.

`&str` is stack-allocated, `String` is heap-allocated; arrays are stack-allocated, vectors are heap-allocated. The preceding chapter refers to `String`s as "growable", this paragraph refers to vectors as growable. That's how I'm reading the analogy. If this isn't right, I'd argue there's something nonintuitive here. IIRC there is is absolutely no other use of the term "slice" in the guides up to this point except for "string slices", so what else can "Vectors are to slices what `String` is to `&str`" possibly be referring to?

File: src/doc/trpl/compound-data-types.md (hunk 4 above)
Comment (ches) [inline src/doc/trpl/compound-data-types.md]: Works for me, I'd agree "below are some examples" is still a little awkward, I was more focused on everything *else* about the original two sentences :smile:  I'm also open to suggestion on a more precise qualifier than "normal".

File: src/doc/trpl/compound-data-types.md (hunk 5 above)
Comment (ches) [inline src/doc/trpl/compound-data-types.md]: Yep, I didn't change it as you can see, but it links to the Generics chapter in the Intermediate section.

File: src/doc/trpl/guessing-game.md (hunk 6 above)
Comment (ches) [inline src/doc/trpl/guessing-game.md]: Sure, seemed about equally divided in source I looked at I think but personally I tend to prefer reference-style too.

File: src/doc/trpl/match.md (hunk 7 above)
Comment (ches) [inline src/doc/trpl/match.md]: I noticed that in one other place, Rust by Example, quite possibly written by the same person :wink: 

Have to ask: is there any strong reasoning or precedent for this choice? I can somewhat understand if there's a desire to differentiate from `switch` statements for programmers with a C background, but "case" is quite the norm for pattern matching in many other languages and in the literature concerning sum types ("case analysis").

File: src/doc/trpl/compound-data-types.md (hunk 2 above)
Comment (ches) [inline src/doc/trpl/compound-data-types.md]: Small matter, but while we're at it: regarding the preference for reference-style linking, it seems not very widely known, but every Markdown parser I've used—including the reference impl I believe—allows this to be written as `[arity]` in the prose instead of `[arity][]` with empty brackets. Is that too magical or surprising? Any opinion?

File: src/doc/trpl/compound-data-types.md (hunk 3 above)
Comment (ches) [inline src/doc/trpl/compound-data-types.md]: Yeah, I was trying to find a concise change to explicitly suggest *why* this could be valuable, in light of the given `Inches(i32)` example ("lets you create a new type that's similar to another one" is kind of just empty, dead verbiage...), but not sure that my edit has quite succeeded. Emphasis was meant to be on "semantic meaning" without any suggestion of behavioral significance. I'll try to come up with a better way to get this across. Maybe the example makes it clear enough without any need to say it, but "similar to another one" still needs to find the chopping block somehow IMO.

File: src/doc/trpl/arrays-vectors-and-slices.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/trpl/arrays-vectors-and-slices.md]: Ahh interesting. The intention with this section is that a vector is a  owning data structure. A slice is a borrowing data structure. A String is an owning data structure. A string slice is a borrowing data structure. Internally, String is a wrapper over a Vec, and &'str is basically the same thing as a slice, a pointer and a length.

File: src/doc/trpl/compound-data-types.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/trpl/compound-data-types.md]: I've been surprised when I've seen it, but maybe because that's just not seen as often. Should be fine. :)

File: src/doc/trpl/guessing-game.md (hunk 6 above)
Comment (steveklabnik) [inline src/doc/trpl/guessing-game.md]: Yeah, I"ve historically preferred the inline style, but the documentation style RFC which recently landed mandates the opposite, so new stuff is written in that style.

File: src/doc/trpl/match.md (hunk 7 above)
Comment (steveklabnik) [inline src/doc/trpl/match.md]: RBE was actually entirely not written by me at all :) Just adopted.

I'm not sure where it comes from historically, but I vaguely remember using 'arms' to describe pattern matching in Haskell, too. They've been called arms as long as I can remember, which is the primary justification, I guess. Maybe an MLism?

**Comment by steveklabnik** [comment]
> Not used to the workflow of the bot,

Yeah, absolutely no worries. Like I said, very happy for the patch in general :)

**Comment by steveklabnik** [comment]
Hey @ches, are you interested in keeping up with this PR?

**Comment by ches** [comment]
Yep! I'm just settling back in after some travel, should be able to update this tomorrow.

What's the preferred protocol on rebasing? I recall reading in contrib guidelines that the team likes follow-up commits to make response to feedback clear. Should I merge master into my topic branch to bring up-to-date?

**Comment by steveklabnik** [comment]
You can feel free to just rebase, it's fine by me. If you go the 'merge master into your topic branch' route, you just end up having to rebase those out at the end anyway.

File: src/doc/trpl/match.md (hunk 7 above)
Comment (ches) [inline src/doc/trpl/match.md]: SML, Haskell, and Erlang use a `case` keyword in pattern matching, in Scala `case` delineates the clauses in a `match` expression; these are a few that I had in mind, along with formal relationship to [case analysis](http://en.wikipedia.org/wiki/Case_analysis#Exhaustive_analysis).

Anyhow, if "arm" has become pervasive in Rust corpus, I'll drop this isolated commit when rebasing.

File: src/doc/trpl/arrays-vectors-and-slices.md (hunk 1 above)
Comment (ches) [inline src/doc/trpl/arrays-vectors-and-slices.md]: We should probably move addressing this part into another patch then—by your clarifying explanation this has several fundamental problems as it stands:

- Ownership hasn't been introduced at all by this stage, it comes later in the intermediate section of the book, so that analogy is completely lost.
- As I noted before, there is is no other use of the term "slice" in the guides up to this point except for "string slice".

   The unqualified term "slice" is introduced later in this chapter, after the analogy line we're discussing. The only logical assumption the reader can make about the analogy at that point is that it refers to string slices, and that makes no sense.

Possibly the part of this chapter that defines slices could be moved ahead of the part on vectors and that would help, although the line that says

> You can also take a slice of a vector, `String`, or `&str`...

should be reworked to not mention vectors before they're defined, but state that slices of vectors are possible later in the vectors part.

This doesn't help with the ownership analogy being lost on the reader, though.

**Comment by ches** [comment]
Rebased on master and updated according to feedback:

- Changed links that I touched to reference-style instead of inline
- Conceding defeat on the term "arm" for pattern matching clauses, for now :stuck_out_tongue_closed_eyes: 
- We lost the guessing game chapter entirely in #22518

I suggest that the vector/slice analogy issue in the very first hidden diff above should be moved to a separate issue. Anything in my list of points in the initial PR message that warrant that too?

**Comment by steveklabnik** [comment]
Thank you!

**Comment by mdinger** [comment]
You write really well. This section just sounds so much better now. You should contribute more often.

**Comment by ches** [comment]
@mdinger Thank you! I'll certainly try. I'm just getting started with Rust in spare time but if there are particular areas of docs that people feel are in need of attention, let me know and I'll look them over with a newcomer's fresh eyes.

**Comment by mdinger** [comment]
Well, there are pain points all over. Follow [reddit/rust](http://www.reddit.com/r/rust/) and they're discussed frequently. Of course there are [doc bugs](https://github.com/rust-lang/rust/labels/A-docs).

Lifetimes questions constantly come up (awesome lifetime docs would be really great!). Partially because there is a lack of clear documentation but also because it's unclear what awesome lifetime docs will look like. See [here](http://users.rust-lang.org/t/what-needs-to-be-explained-about-lifetime-parameters/291/2) for examples and links to perplexed readers.

Closures were a problem (mainly outdated because closures just changed a lot) but it looks like @steveklabnik is about to make them a bit better with https://github.com/rust-lang/rust/pull/23568 .

Making one cohesive document which handles people of all backgrounds proves problematic. Systems programmers don't like the simple things explained. Non-systems guys don't them glossed. Deep type language tends to be used by developers which can cause problems until simpler docs are written: see https://github.com/rust-lang/rust/issues/22914

Docs are still a moving target, but it's getting better. Compiler breaking old Rust is decreasing and this makes things a lot easier. Also check out [rbe](http://rustbyexample.com/) if you haven't seen it. It's not prominently advertised on the main site.
